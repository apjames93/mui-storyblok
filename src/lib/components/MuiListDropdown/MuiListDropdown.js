import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { List, Collapse } from '@material-ui/core';
import MuiListExpansion from './components/MuiListExpansion/MuiListExpansion';

import MuiList from '../MuiList/MuiList';
import StoryBlok from '../../utils/Storyblok';
import { validComponentsRequired, validComponents } from '../../utils/customProps';
import { renderComponents } from '../../utils/customComponents';

export const MuiListDropdown = ({
  rootClass,
  listExpansion,
  content,
}) => {
  const styles = StoryBlok.arrayToMuiStyles(rootClass);
  const [open, setOpen] = useState(false);
  const handleClick = () => {
    setOpen(!open);
  };

  return (
    <List className={styles.root}>
      <MuiListExpansion {...listExpansion[0]} handleClick={handleClick} open={open} />
      <Collapse in={open} timeout="auto" unmountOnExit>
        {content.map((component, key) => renderComponents({ MuiList }, component, key))}
      </Collapse>
    </List>
  );
};

export default MuiListDropdown;

MuiListDropdown.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Override or extend the styles applied to the component
   * */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * components:
   * MuiListItem
   */
  content(props, propName, componentName) {
    const components = ['MuiList'];
    return validComponents(props, propName, componentName, components);
  },
  /**
   * components:
   * MuiListItem
   */
  listExpansion(props, propName, componentName) {
    const components = ['MuiListItem'];
    return validComponentsRequired(props, propName, componentName, components, 1);
  },
};

MuiListDropdown.defaultProps = {
  rootClass: [],
  content: [],
  listExpansion: [],
};
