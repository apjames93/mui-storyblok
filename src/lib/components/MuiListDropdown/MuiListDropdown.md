##### MuiListDropdown has the ability to render custom components.