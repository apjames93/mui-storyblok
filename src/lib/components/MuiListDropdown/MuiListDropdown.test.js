import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import MuiListDropdown from './MuiListDropdown';

function setup() {
  const props = {
    listExpansion: [{
      component: 'MuiListExpansion',
      icon: [{
        component: 'MuiListItemIcon',
        content: [{
          component: 'MuiIcon',
          iconName: 'star',
        }],
      }],
      primaryText: 'list Text',
    }],
    content: [{
      component: 'MuiList',
      content: [{
        component: 'MuiListItem',
      }],
    }],
  };
  const comp = shallow(<MuiListDropdown {...props} />);
  return { comp, props };
}

describe('<MuiListDropdown />', () => {
  it('renders MuiListDropdown', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<MuiListDropdown {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should handle click and toggle open prop', () => {
    const { comp } = setup();
    const initialOpen = comp.find('MuiListExpansion').first().prop('open');
    expect(initialOpen).toEqual(false);
    const listExpansion = comp.find('MuiListExpansion').first().prop('handleClick');
    listExpansion();
    const isOpen = comp.find('MuiListExpansion').first().prop('open');
    expect(isOpen).toEqual(true);
  });
});
