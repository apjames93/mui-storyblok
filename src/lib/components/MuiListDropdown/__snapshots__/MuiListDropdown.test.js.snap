// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiListDropdown /> snapshot 1`] = `
<ul
  className="MuiList-root makeStyles-root-3 MuiList-padding"
>
  <div
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root makeStyles-root-4 MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemIcon-root makeStyles-root-5"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root makeStyles-root-6 MuiIcon-colorSecondary"
      >
        star
      </span>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        list Text
      </span>
    </div>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
  </div>
</ul>
`;
